### PR TITLE
Handle different type property name case

### LINKED ISSUE #31: Support for both camel case and non camel case parameters
Here's my current setup:

  ```
  [JsonConverter(typeof(JsonSubtypes), "Type")]
    [JsonSubtypes.KnownSubType(typeof(A), PriceDiscountType.A)]
    [JsonSubtypes.KnownSubType(typeof(B), PriceDiscountType.B)]
    [JsonSubtypes.KnownSubType(typeof(C), PriceDiscountType.C)]
    [JsonSubtypes.KnownSubType(typeof(D), PriceDiscountType.D)]
    [JsonSubtypes.KnownSubType(typeof(E), PriceDiscountType.E)]
    [JsonSubtypes.KnownSubType(typeof(F), PriceDiscountType.F)]
    [JsonSubtypes.KnownSubType(typeof(G), PriceDiscountType.G)]
    public abstract class PriceDiscount : Core.Toolkit.ICloneable
```

However, at times the default serialization is to use camel case.  Is there a way to make this work for both camel case, and non camel case, so that I don't have to worry about which is being used?

### LINKED ISSUE #38: JsonConverter doesn't respect naming rules
Code sample
```cs
public class DtoTest
{
	[Fact]
	public void FooParsing()
	{
		CheckDeserialization<Foo>("{\"msgType\":\"1\"}");
	}

	private static void CheckDeserialization<T>(string json) where T : class 
	{
		var result = JsonConvert.DeserializeObject<DtoBase>(json);
		Assert.IsType<T>(result);
	}
}

[JsonConverter(typeof(JsonSubtypes), nameof(DtoBase.MsgType))]
[JsonSubtypes.KnownSubType(typeof(Foo), 1)]
public abstract class DtoBase
{
	public virtual int MsgType { get; }
}

class Foo : DtoBase
{
	public override int MsgType { get; } = 1;
}
```

### Expected behaviour

Test passed

### Actual behaviour

> Newtonsoft.Json.JsonSerializationException : Could not create an instance of type Sample.Test.DtoBase. Type is an interface or abstract class and cannot be instantiated. Path 'msgType', line 1, position 11.
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateNewObject(JsonReader reader, JsonObjectContract objectContract, JsonProperty containerMember, JsonProperty containerProperty, String id, Boolean& createdFromNonDefaultCreator)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
>    at JsonSubTypes.JsonSubtypes.ThreadStaticReadObject(JsonReader reader, JsonSerializer serializer, JToken jToken, Type targetType)
>    at JsonSubTypes.JsonSubtypes.ReadJson(JsonReader reader, Type objectType, JsonSerializer serializer)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.DeserializeConvertable(JsonConverter converter, JsonReader reader, Type objectType, Object existingValue)
>    at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
>    at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
>    at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
>    at Sample.Test.DtoTest.CheckDeserialization[T](String json) in C:\Users\Alex\Documents\Repo\SampleEthereum\Sample.Test\UnitTest1.cs:line 17
>    at Sample.Test.DtoTest.FooParsing() in C:\Users\Alex\Documents\Repo\SampleEthereum\Sample.Test\UnitTest1.cs:line 12

---

Test passes for `[JsonConverter(typeof(JsonSubtypes), "msgType")]` attribute value. camelCase is hardcoded somewhere.

### PR DESCRIPTION
**Pull request by manuc66**
Now handle different type property name case 
resolve #31 
resolve #38

### CODE REVIEW

File: JsonSubTypes.Tests/DemoKnownSubTypeWithProperty.cs [hunk 1]
```diff
@@ -40,6 +40,19 @@ public void Demo()
...
+        public void DemoDifferentCase()
+        {
+            string json = "[{\"Department\":\"Department1\",\"JobTitle\":\"JobTitle1\",\"FirstName\":\"FirstName1\",\"LastName\":\"LastName1\"}," +
+                          "{\"Department\":\"Department1\",\"JobTitle\":\"JobTitle1\",\"FirstName\":\"FirstName1\",\"LastName\":\"LastName1\"}," +
+                          "{\"skill\"" +
+                          ":\"Painter\",\"FirstName\":\"FirstName1\",\"LastName\":\"LastName1\"}]";
+
+
+            var persons = JsonConvert.DeserializeObject<ICollection<Person>>(json);
+            Assert.AreEqual("Painter", (persons.Last() as Artist)?.Skill);
```
Comment (Pzixel) [inline JsonSubTypes.Tests/DemoKnownSubTypeWithProperty.cs]: I'm suggesting to add yet another Artist with `Skill` to show that both cases are valid

**Comment by Pzixel** [comment]
Fixes #38

File: JsonSubTypes.Tests/DemoKnownSubTypeWithProperty.cs (hunk 1 above)
Comment (manuc66) [reply JsonSubTypes.Tests/DemoKnownSubTypeWithProperty.cs]: It's already there on [line 34](https://github.com/manuc66/JsonSubTypes/pull/39/files/493d48564587cd9878a02154fbe4d543d790261f#diff-bb2cbc8ae9988a0a27793302e2dd0d64R34)

Comment (Pzixel) [reply JsonSubTypes.Tests/DemoKnownSubTypeWithProperty.cs]: Ah ok, I have only seen the diff.

**Comment by manuc66** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L74](https://github.com/manuc66/JsonSubTypes/blob/8b225deb67bdcd2e6740c97f9fb29dafa60ac5ea/JsonSubTypes/JsonSubtypesConverter.cs#L74): Remove this set, which updates a 'static' field from an instance property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)
1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L86](https://github.com/manuc66/JsonSubTypes/blob/8b225deb67bdcd2e6740c97f9fb29dafa60ac5ea/JsonSubTypes/JsonSubtypesConverter.cs#L86): Remove this set, which updates a 'static' field from an instance method. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)
1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L87](https://github.com/manuc66/JsonSubTypes/blob/8b225deb67bdcd2e6740c97f9fb29dafa60ac5ea/JsonSubTypes/JsonSubtypesConverter.cs#L87): Remove this set, which updates a 'static' field from an instance method. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Comment by manuc66** [comment]
SonarQube analysis reported 3 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L74](https://github.com/manuc66/JsonSubTypes/blob/95acc66cb5f6e6d3e959e517c5c96a55ec64d386/JsonSubTypes/JsonSubtypesConverter.cs#L74): Remove this set, which updates a 'static' field from an instance property. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)
1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L86](https://github.com/manuc66/JsonSubTypes/blob/95acc66cb5f6e6d3e959e517c5c96a55ec64d386/JsonSubTypes/JsonSubtypesConverter.cs#L86): Remove this set, which updates a 'static' field from an instance method. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)
1. ![CRITICAL][CRITICAL] [JsonSubtypesConverter.cs#L87](https://github.com/manuc66/JsonSubTypes/blob/95acc66cb5f6e6d3e959e517c5c96a55ec64d386/JsonSubTypes/JsonSubtypesConverter.cs#L87): Remove this set, which updates a 'static' field from an instance method. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS2696)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'

**Comment by Pzixel** [comment]
Well, Sonar is right, it's not good idea to update static fields from instance methods.

**Comment by manuc66** [comment]
@Pzixel I would agree but those fields are thread static and it's the only "hack" that I could find to transmit information through stack calls without modifying JSON.NET. And even if it's considered bad practice in general in this particular case it's safe.

If you have a better idea you're really welcome ;-)

**Comment by Pzixel** [comment]
@manuc66 why not just do it instance method? In this case it should't be threadlocal. If you consider single instance accessing from different threads just use `Interlocked` to update values.

**Comment by manuc66** [comment]
@Pzixel it's per thread instance

**Comment by Pzixel** [comment]
@manuc66 why should one instance update this value for another one?

**Comment by manuc66** [comment]
@Pzixel static field annotated with thread local means that each thread [has it own value](https://msdn.microsoft.com/en-us/library/system.threadstaticattribute(v=vs.110).aspx) and this library "remember" that the converter has already been called one time on a thread so that it could delegate the current and subtree de-serialisation to JSON.NET once the subtype information has been discovered.

**Comment by manuc66** [comment]
@Pzixel it shouldnt need but should not produce nasty effect since the purpose is to skip the second call to converter once it has already been called and in between there is no possibility to be in another converter instance.

**Comment by Pzixel** [comment]
@manuc66 I propose to not introduce premature optimization until it's really required.

**Comment by manuc66** [comment]
@Pzixel I I assure you that the use of thread static has really not been taken inconsiderately from my side. If you think you can solve the problematic with a cleaner approach feel free to submit a PR.

**Comment by Pzixel** [comment]
@manuc66 if you have some benchmark that shows that it really worth it then I have to arguments why it shouldn't be done this way. Otherwise I suggest either add them or just make it instance fields and check if it works fine.

I'd like to suggest some cleaner solution for the former case too, but I don't really get the whole requirements set yet.

**Comment by manuc66** [comment]
@Pzixel In the benchmark are in the test suite, if you could solve it without breaking the test suite you got it. Currently if you remove `[ThreadStatic] static` qualifier you have 3 failing tests.

**Comment by Pzixel** [comment]
Well, if you are this sure about the solution then this warning should be ignored via some attribute or something, I guress.

**Comment by manuc66** [comment]
@Pzixel The warning popout in pull request but are marked as wont fix/false positive on sonar cloud for the master branch see: https://sonarcloud.io/project/issues?id=manuc66%3AJsonSubtypes%3Amaster&resolutions=WONTFIX&types=CODE_SMELL

**Pull request merged**